jdaviz/configs/imviz/plugins/orientation/orientation.py

```diff
@@ -351,36 +354,54 @@ def add_orientation(self, rotation_angle=None, east_left=None, label=None,
         )
 
         # add orientation layer to all viewers:
-        self._add_data_to_all_viewers(label)
+        for viewer_ref in self.app._viewer_store:
+            self._add_data_to_viewer(label, viewer_ref)
 
         if set_on_create:
-            self.orientation._update_layer_items()
             self.orientation.selected = label
 
-    def _add_data_to_all_viewers(self, data_label):
-        for viewer_ref in self.app.get_viewer_reference_names():
-            layers = [
-                layer.label for layer in
-                self.app.get_viewer(viewer_ref).layers
-            ]
-            if data_label not in layers:
-                self.app.add_data_to_viewer(viewer_ref, data_label)
+    def _add_data_to_viewer(self, data_label, viewer_id):
+        viewer = self.app.get_viewer_by_id(viewer_id)
+
+        wcs_only_layers = viewer.state.wcs_only_layers
+        if data_label not in wcs_only_layers:
+            self.app.add_data_to_viewer(viewer_id, data_label)
 
     def _on_viewer_added(self, msg):
-        for data_label in self.orientation.choices:
-            self._add_data_to_all_viewers(data_label)
+        self._send_wcs_layers_to_all_viewers(viewers_to_update=[msg._viewer_id])
+
+    @observe('viewer_items')
+    def _send_wcs_layers_to_all_viewers(self, *args, **kwargs):
+        if not hasattr(self, 'viewer'):
+            return
+
+        wcs_only_layers = self.app.get_viewer_by_id('imviz-0').state.wcs_only_layers
+
+        viewers_to_update = kwargs.get(
+            'viewers_to_update', self.app.get_viewer_reference_names()
+        )
 
-        if hasattr(self, 'orientation') and len(self.orientation.choices):
-            self.viewer.selected = msg._viewer_id
-            self.orientation.selected = self.orientation.choices[0]
+        for viewer_ref in viewers_to_update:
+            for wcs_layer in wcs_only_layers:
+                if wcs_layer not in self.app.get_viewer_by_id(viewer_ref).state.wcs_only_layers:
+                    self.app.add_data_to_viewer(viewer_ref, wcs_layer)
+
+            self.viewer.selected = viewer_ref

```

Comment:
this will actually cycle the viewer select in the orientation plugin, right?  Is that for the sake of having the choices update in `self.orientation`?  Why do we need to do that now for all `viewers_to_update`?

Reply:
We need that because we call `self.orientation.update_wcs_only_filter` below, which is particular to each viewer.

Reply:
but `self.orientation` won't remember per-viewer filters, so setting for anything but the last iteration of the for-loop will just be wasted effort... I think

Reply:
The one time that this is necessary is when a new viewer is created and the WCS-only reference data have been added to the viewer but not yet selected as reference data. We first need to change the filter to "see" the orientation options, then we need to select the default orientation – I do this a few lines below (https://github.com/bmorris3/jdaviz/pull/29/files/e2f36a547e94cb0d38e45b807cb7dfba38c85122#diff-022bcba39f3fd68183229b5433146c2a21ac800db2778790152edc9a6b1b01caR398).

Reply:
ok, and in that case `viewers_to_update` is passed and only has one entry it seems.  I wonder if instead of having 

```
self.viewer.selected = viewer_ref
...
self.orientation.selected = base_wcs_layer_label
```

in the for-loop and run for all (passed) viewers, we should instead have 

```
if self.viewer.selected == viewer:
    ....
    self.orientation.selected = base_wcs_layer_label
```

Does that make sense?

Reply:
I understand what you're going for, but this method is _also_ used when new orientation layers are created. In that case, it adds the new orientation layer to each viewer, so that it can be selected as refdata in that viewer. So the function is written for the general case.